rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -57,93 +69,240 @@ def _maybe_decode_v3_swap(  # pylint: disable=no-self-use
-                    # In this branch of the condition we also add the transfer to correctly decode
-                    # the case where ETH was sent to the user in a multi-step swap that was decoded
-                    # by the event in the previous step. The check for the location label and the
-                    # amount ensure that we are decoding the right event.
-                    event.event_type = HistoryEventType.TRADE
-                    event.event_subtype = HistoryEventSubType.RECEIVE
-                    event.counterparty = CPT_UNISWAP_V3
-                    event.notes = f'Receive {event.balance.amount} {crypto_asset.symbol} in uniswap-v3 from {event.location_label}'  # noqa: E501
-                    in_event = event
-                elif (
-                    event.event_type == HistoryEventType.RECEIVE and
-                    event.balance.amount != asset_normalized_value(
-                        amount=amount_received,
-                        asset=crypto_asset,
-                    ) and event.counterparty == UNISWAP_V3_ROUTER
-                ):
-                    # this is to make sure its the amm issuing the refund and not an agreggator making a swap  # noqa: E501
-                    # Those are assets returned due to a change in the swap price
-                    event.event_type = HistoryEventType.TRANSFER
-                    event.event_subtype = HistoryEventSubType.NONE
-                    event.counterparty = CPT_UNISWAP_V3
-                    event.notes = f'Refund of {event.balance.amount} {crypto_asset.symbol} in uniswap-v3 due to price change'  # noqa: E501
-
-        maybe_reshuffle_events(out_event=out_event, in_event=in_event)
+        if transaction.to_address not in UNISWAP_ROUTERS:
+            return decoded_events  # work only with the known routers for now
+
+        if len(decoded_events) < 3 or decoded_events[0].counterparty != CPT_GAS:
+            # The minimum is 3 events: gas, spend, receive.
+            return decoded_events
+
+        # Since we check that the to_address is a known router, we can be sure that the ETH
+        # transfer event (and other events) are parts of the swap.
+        if decoded_events[1].asset == A_ETH:
+            if decoded_events[1].event_type == HistoryEventType.SPEND:
+                # This is a swap from eth to token
+                swap_data = self._decode_eth_to_token_swap(decoded_events)
+            else:  # Receive
+                # This is a swap from token to eth
+                swap_data = self._decode_token_to_eth_swap(decoded_events)
+        else:
+            # Then this should be a swap from token to token
+            swap_data = self._decode_token_to_token_swap(decoded_events)
+
+        if swap_data is None or swap_data.from_asset is None or swap_data.to_asset is None:
+            self.msg_aggregator.add_warning(
+                f'Failed to decode a uniswap swap for transaction {transaction.tx_hash.hex()}. '
```

Comment:
Isn't this kind of message a bit verbose? Do we do this for transactions that we have errors with? Also what is the difference between this and the self.notify_user() method? Just that one calls add_warning and the other add_error?

Reply:
The thing is that `notify_user` method is only for assets. The text that it sends is:
```
f'Could not identify asset {event.asset} decoding ethereum event in {counterparty}. '
f'Make sure that it has all the required properties (name, symbol and decimals) and '
f'try to decode the event again {event.event_identifier.hex()}.',
```

But here there is a different problem

Reply:
Discussed in the standup. Will change into a log.error